snappy_pipeline/workflows/somatic_neoepitope_prediction/__init__.py

```diff
@@ -31,11 +31,338 @@
+                              # the column containing expression data.
+                              # Required when using the `custom` and `star` format. For `star`, there are
+                              # only 3 options [unstranded,rf,fr]
+      ignore-ensembl-id-version: True #Ignore the ensemble id version
 """
+
+
+class NeoepitopePreparationStepPart(BaseStepPart):
+    """
+    Preparation VCF file for pvactool
+    """
+
+    #: Step name
+    name = "neoepitope_preparation"
+    actions = ("run",)
+
+    def __init__(self, parent):
+        super().__init__(parent)
+        # Build shortcut from cancer bio sample name to matched cancer sample
+        self.tumor_ngs_library_to_sample_pair = OrderedDict()
+        for sheet in self.parent.shortcut_sheets:
+            self.tumor_ngs_library_to_sample_pair.update(
+                sheet.all_sample_pairs_by_tumor_dna_ngs_library
+            )
+        # Build mapping from donor name to donor.
+        self.donors = OrderedDict()
+        for sheet in self.parent.shortcut_sheets:
+            for donor in sheet.donors:
+                self.donors[donor.name] = donor
+
+    @dictify
+    def get_input_files(self, action):
+        """Return path to somatic variant annotated file"""
+        # It only works with vep now.
+        # Validate action
+        self._validate_action(action)
+        tpl = (
+            "output/{mapper}.{var_caller}.{anno_caller}.{tumor_library}/out/"
+            "{mapper}.{var_caller}.{anno_caller}.{tumor_library}"
+        )
+        # Need to change for work on many different tools
+        key_ext = {"vcf": ".full.vcf.gz", "vcf_tbi": ".full.vcf.gz.tbi"}
+        variant_annotation = self.parent.sub_workflows["somatic_variant_annotation"]
+        ngs_mapping = self.parent.sub_workflows["ngs_mapping"]
+        for key, ext in key_ext.items():
+            yield key, variant_annotation(tpl + ext)
+        # Getting appropriate rna library
+        for sheet in filter(is_not_background, self.parent.sheets):
+            for donor in sheet.bio_entities.values():
+                for biosample in donor.bio_samples.values():
+                    if biosample.extra_infos["isTumor"]:
+                        for test_sample in biosample.test_samples.values():
+                            if test_sample.extra_infos["extractionType"] == "RNA":
+                                for lib in test_sample.ngs_libraries.values():
+                                    rna_library = lib.name
+                                    rna_tpl = (
+                                        "output/{mapper}.{library_name}/out/{mapper}.{library_name}"
+                                    ).format(
+                                        mapper="star",
+                                        library_name=rna_library,
+                                    )
+                                    ext = {"expression", "bam", "bai"}
```

Comment:
The nesting is also pretty deep, perhaps it can be restructured (by breaking it into smaller chunks / moving common stuff into dedicated functions) a bit to make it a bit more palatable